loopd/daemon.go

```diff
@@ -161,75 +184,191 @@ func daemon(config *config, lisCfg *listenerCfg) error {
 		log.Infof("REST proxy disabled")
 	}
 
-	mainCtx, cancel := context.WithCancel(context.Background())
-	var wg sync.WaitGroup
-
-	// Start the swap client itself.
-	wg.Add(1)
+	// Start the grpc server.
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
-		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, server.statusChan)
+		log.Infof("RPC server listening on %s", d.grpcListener.Addr())
+		err = d.grpcServer.Serve(d.grpcListener)
 		if err != nil {
 			log.Error(err)
 		}
-		log.Infof("Swap client stopped")
+	}()
 
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
+	return nil
+}
 
-		cancel()
-	}()
+// StartAsSubserver is an alternative to Start where the RPC server does not
+// create its own gRPC server but registers to an existing one. The same goes
+// for REST (if enabled), instead of creating an own mux and HTTP server, we
+// register to an existing one.
+func (d *Daemon) StartAsSubserver(lndGrpc *lndclient.GrpcLndServices) error {
+	// There should be no reason to start the daemon twice. Therefore return
+	// an error if that's tried. This is mostly to guard against Start and
+	// StartAsSubserver both being called.
+	if atomic.AddInt32(&d.started, 1) != 1 {
+		return errOnlyStartOnce
+	}
 
-	// Start a goroutine that broadcasts swap updates to clients.
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
+	// When starting as a subserver, we get passed in an already established
+	// connection to lnd that might be shared among other subservers.
+	d.lnd = lndGrpc
 
-		log.Infof("Waiting for updates")
-		server.processStatusUpdates(mainCtx)
-	}()
+	// With lnd already pre-connected, initialize everything else, such as
+	// the swap server client, the RPC server instance and our main swap
+	// handlers.
+	err := d.initialize()
+	if err != nil {
+		return err
+	}
 
-	// Start the grpc server.
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
+	return nil
+}
```

Comment:
Discussed offline: what happens if something goes wrong between `Start` and `Stop`

Reply:
Addressed with `errChan`.